src/.clang-format

```
@@ -1,50 +1,96 @@
+# -*- mode: yaml; -*-
+
+# Goals
+# - have automated formatting to avoid wasted time
+# - minimize diff-sensitivity to things like changing function names (e.g., don’t want whitespace
+#   diffs on all the parameter lines)
+
+# NB: If clang-format can’t parse this file (or doesn’t understand any of the values, it will format
+#     with the default settings. If this happens it’s important to _revert_ the formatting, not just
+#     reformat after fixing it. This is because some formatting changes aren’t reversible, so you
+#     may introduce formatting regressions if you simply reformat.
+
+# TODO: clang-format 17 adds the ability to include a trailing comment on formatting delimiters, so
+#       once we’re using that version
+#
+#     // NB: why we’re disabling
+#     // clang-format off
+#
+#       can become
+#
+#     // clang-format off: why we’re disabling
+
 AccessModifierOffset: -4
-AlignEscapedNewlinesLeft: true
+AlignAfterOpenBracket: AlwaysBreak
+AlignEscapedNewlines: Left
 AlignTrailingComments: true
+AllowAllArgumentsOnNextLine: false
+AllowAllConstructorInitializersOnNextLine: true
 AllowAllParametersOfDeclarationOnNextLine: false
-AllowShortBlocksOnASingleLine: false
-AllowShortFunctionsOnASingleLine: All
+AllowShortBlocksOnASingleLine: Empty
+AllowShortCaseLabelsOnASingleLine: true
+AllowShortFunctionsOnASingleLine: Inline
 AllowShortIfStatementsOnASingleLine: false
+AllowShortLambdasOnASingleLine: All
 AllowShortLoopsOnASingleLine: false
 AlwaysBreakBeforeMultilineStrings: false
 AlwaysBreakTemplateDeclarations: true
+BinPackArguments: false
 BinPackParameters: false
-BreakBeforeBinaryOperators: false
+BreakBeforeBinaryOperators: NonAssignment
 BreakBeforeBraces: Linux
-BreakBeforeTernaryOperators: false
-BreakConstructorInitializersBeforeComma: false
-ColumnLimit:     0
-CommentPragmas:  '^ IWYU pragma:'
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializers: BeforeColon
+ColumnLimit: 100 # match rustfmt, if we don’t define this, then we don’t get Wadler
+CommentPragmas: '^ IWYU pragma:'
 ConstructorInitializerAllOnOneLineOrOnePerLine: false
 ConstructorInitializerIndentWidth: 4
 ContinuationIndentWidth: 4
 Cpp11BracedListStyle: true
 DerivePointerAlignment: false
-DisableFormat:   false
-IndentCaseLabels: false
+DisableFormat: false
+IncludeBlocks: Regroup
+IncludeCategories:
+    - Regex:           '<[^/.]+>'
+      Priority:        3
+    - Regex:           '<.*>'
+      Priority:        2
+    - Regex:           '.*'
+      Priority:        1
```

`#include`s should *never* be automatically reordered by a formatter; I am adamant about that. Unlike reordering Rust imports, reordering `#include`s can change behaviour.
```suggestion
SortIncludes: false
```